dll/win32/kernel32/client/sysinfo.c

```diff
@@ -413,68 +414,238 @@ GetNumaAvailableMemoryNode(IN UCHAR Node,
+        return 0;
+    }
+
+    /* Call unicode version interface */
+    Length = GetFirmwareEnvironmentVariableExW(VariableName.Buffer,
+                                               Namespace.Buffer,
+                                               pBuffer,
+                                               nSize,
+                                               pdwAttribubutes);
+
+    /* Clean and return */
+    RtlFreeUnicodeString(&Namespace);
+    RtlFreeUnicodeString(&VariableName);
+    return Length;
 }
 
-/*
- * @unimplemented
- */
+_Success_(return > 0)
 DWORD
 WINAPI
-GetFirmwareEnvironmentVariableA(IN LPCSTR lpName,
-                                IN LPCSTR lpGuid,
-                                IN PVOID pValue,
-                                IN DWORD nSize)
+GetFirmwareEnvironmentVariableW(
+    _In_ LPCWSTR lpName,
+    _In_ LPCWSTR lpGuid,
+    _Out_writes_bytes_to_opt_(nSize, return) PVOID pBuffer,
+    _In_ DWORD nSize)
 {
-    STUB;
-    return 0;
+    return GetFirmwareEnvironmentVariableExW(lpName, lpGuid, pBuffer, nSize, NULL);
+}
+
+_Success_(return > 0)
+DWORD
+WINAPI
+GetFirmwareEnvironmentVariableA(
+    _In_ LPCSTR lpName,
+    _In_ LPCSTR lpGuid,
+    _Out_writes_bytes_to_opt_(nSize, return) PVOID pBuffer,
+    _In_ DWORD nSize)
+{
+    return GetFirmwareEnvironmentVariableExA(lpName, lpGuid, pBuffer, nSize, NULL);
 }
 
-/*
- * @unimplemented
- */
 BOOL
 WINAPI
-SetFirmwareEnvironmentVariableA(IN LPCSTR lpName,
-                                IN LPCSTR lpGuid,
-                                IN PVOID pValue,
-                                IN DWORD nSize)
+SetFirmwareEnvironmentVariableExW(
+    _In_ LPCWSTR lpName,
+    _In_ LPCWSTR lpGuid,
+    _In_reads_bytes_opt_(nSize) PVOID pValue,
+    _In_ DWORD nSize,
+    _In_ DWORD dwAttributes)
 {
-    STUB;
-    return 0;
+    NTSTATUS Status;
+    UNICODE_STRING VariableName, Namespace;
+    GUID NamespaceGuid;
+
+    /* Check input parameters and build NT strings */
+    if (!lpName || !lpGuid)
+    {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+    RtlInitUnicodeString(&VariableName, lpName);
+    RtlInitUnicodeString(&Namespace, lpGuid);
+    Status = RtlGUIDFromString(&Namespace, &NamespaceGuid);
+    if (!NT_SUCCESS(Status))
+    {
+        BaseSetLastNTError(Status);
+        return FALSE;
+    }
+
+    /* Set firmware system environment variable value */
+    Status = NtSetSystemEnvironmentValueEx(&VariableName, &NamespaceGuid, pValue, nSize, dwAttributes);
+    if (!NT_SUCCESS(Status))
+    {
+        BaseSetLastNTError(Status);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+BOOL
+WINAPI
+SetFirmwareEnvironmentVariableExA(
+    _In_ LPCSTR lpName,
+    _In_ LPCSTR lpGuid,
+    _In_reads_bytes_opt_(nSize) PVOID pValue,
+    _In_ DWORD nSize,
+    _In_ DWORD dwAttributes)
+{
+    NTSTATUS Status;
+    BOOL Result;
+    UNICODE_STRING VariableName, Namespace;
+    ANSI_STRING AnsiVariableName, AnsiNamespaceGuid;
```

Comment:
Maybe you can use shorter variable names, like "Name, Guid" and "AnsiName, AnsiGuid".

Reply:
Why?

Reply:
I suggest use `AnsiVariableName` and `AnsiNamespace`, won't be so long